graph/src/components/metrics/stopwatch.rs

```diff
@@ -0,0 +1,169 @@
+    }
+}
+
+/// Usage example:
+/// ```ignore
+/// // Start counting time for the "main_section".
+/// let _main_section = stopwatch.start_section("main_section");
+/// // do stuff...
+/// // Pause timer for "main_section", start for "child_section".
+/// let chid_section = stopwatch.start_section("child_section");
+/// // do stuff...
+/// // Register time spent in "child_section", implicitly going back to "main_section".
+/// section.end();
+/// // do stuff...
+/// // At the end of the scope `_main_section` is dropped, which is equivalent to calling
+/// // `_main_section.end()`.
+#[derive(Clone)]
+pub struct StopwatchMetrics {
+    disabled: Arc<AtomicBool>,
+    inner: Arc<Mutex<StopwatchInner>>,
+}
+
+impl StopwatchMetrics {
+    pub fn new(
+        logger: Logger,
+        subgraph_id: SubgraphDeploymentId,
+        registry: Arc<dyn MetricsRegistry>,
+    ) -> Self {
+        let mut inner = StopwatchInner {
+            total_counter: *registry
+                .new_counter(
+                    format!("{}_total_secs", subgraph_id),
+                    format!("total time spent syncing"),
+                    HashMap::new(),
+                )
+                .expect("failed to register total_secs prometheus counter"),
+            logger,
+            subgraph_id,
+            registry,
+            counters: HashMap::new(),
+            section_stack: Vec::new(),
+            timer: Instant::now(),
+        };
+
+        // Start a base section so that all time is accounted for.
+        inner.start_section("unknown".to_owned());
+
+        StopwatchMetrics {
+            disabled: Arc::new(AtomicBool::new(false)),
+            inner: Arc::new(Mutex::new(inner)),
+        }
+    }
+
+    pub fn start_section(&self, id: &str) -> Section {
+        let id = id.to_owned();
+        if !self.disabled.load(Ordering::SeqCst) {
+            self.inner.lock().unwrap().start_section(id.clone())
+        }
+        Section {
+            id,
+            stopwatch: self.clone(),
+        }
+    }
+
+    pub fn disable(&self) {
+        self.disabled.store(true, Ordering::SeqCst)
+    }
+
+    fn end_section(&self, id: String) {
+        if !self.disabled.load(Ordering::SeqCst) {
+            self.inner.lock().unwrap().end_section(id)
+        }
+    }
+}
+
+/// We want to account for all subgraph indexing time, based on "wall clock" time. To do this we
+/// break down indexing into _sequential_ sections, and register the total time spent in each. So
+/// that there is no double counting, time spent in child sections doesn't count for the parent.
+struct StopwatchInner {
+    logger: Logger,
+    subgraph_id: SubgraphDeploymentId,
+    registry: Arc<dyn MetricsRegistry>,
+
+    // Counter for the total time the subgraph spent syncing.
+    total_counter: Counter,
+
+    // Counts the seconds spent in each section of the indexing code.
+    counters: HashMap<String, Counter>,
+
+    // The top section is the one that's currently executing.
```

Comment:
By `top`, do you mean the last section in the vector or the first?

Reply:
The top of the stack is the last element, I've added a comment.